src/hammer-vlsi/hammer_vlsi/hammer_tool.py

```diff
@@ -1064,6 +1064,206 @@ def get_bumps(self) -> Optional[BumpsDefinition]:
             pitch=Decimal(str(self.get_setting("vlsi.inputs.bumps.pitch"))),
             cell=self.get_setting("vlsi.inputs.bumps.cell"), assignments=assignments)
 
+    def generate_visualization(self) -> None:
+        """
+        Generate an SVG visualizing the chip.
+        Depending on mode, it will display floorplan (placement constraints for the current hierarchy)
+        and/or bumps (overlaid on top of floorplan) and pad designators (if this tool is a PCBDeliverableTool).
+        Call this from any custom hook (not used by any default flow).
+        Note: visualizations generated within a PCBDeliverableTool are mirrored about the y-axis (assumption: flip-chip on PCB)
+        """
+        viz_mode = self.get_setting("vlsi.inputs.visualization.mode") #type: str
+        viz_file = self.get_setting("vlsi.inputs.visualization.svg_file") #type: str
+        shorten_path_depth = self.get_setting("vlsi.inputs.visualization.shorten_path_depth") #type: int
+
+        # Checks
+        if not viz_mode in ["all", "floorplan", "bumps", "footprint"]:
+            self.logger.error("Invalid type for 'vlsi.inputs.visualization.type'! No visualization generated.")
+            return
+
+        fp_consts = self.get_placement_constraints() #type: List[PlacementConstraint]
+        # TODO: make pcb action recognize toplevel constraint if par action had hierarchical constraints
+        try:
+            # Get toplevel constraint
+            top = next(filter(lambda x: x.type == PlacementConstraintType.TopLevel, fp_consts)) #type: PlacementConstraint
+        except:
+            self.logger.error("You must at least have a 'toplevel' type placement constraint to generate a visualization!")
+            return
+
+        # TODO: will break if an alternate PCB tool is implemented without naming_scheme property.
+        # TODO: support fractional bump coordinates when generating footprint
+        from .hammer_vlsi_impl import HammerPCBDeliverableTool
+        is_pcb_tool = isinstance(self, HammerPCBDeliverableTool)
+        if not is_pcb_tool:
+            if viz_mode == "footprint":
+                self.logger.error("Can't generate footprint alone if this isn't a PCBDeliverableTool.")
+                return
+            elif viz_mode == "all":
+                self.logger.info("Visualizing in 'all' mode but not a PCBDeliverableTool. Skipping drawing pad designators.")
+
+        # Shorten path names to first letters except lowest "depth" levels in hierarchy
+        def shorten(path: str) -> str:
+            if shorten_path_depth > 0:
+                parent = path.split('/')[:-shorten_path_depth]
+                inst = path.split('/')[-shorten_path_depth:]
+                return os.path.join(*[p[0] for p in parent], *inst)
```

Comment:
I was expecting this to be a regular python join not path join so was confused why it didn't include `/`.
For reference I think the equivalent "normal" join code is:
```
"/".join([p[0] for p in parent] + inst)
```
I don't really care one way or the other it just seems like a strange use of the `os.path` stuff

Reply:
Yeah, the way I did it was definitely a case of trying to reduce some lines of code and not realizing I could do a simple string join...